### PR TITLE
Add support for sandwiches

### PR DESCRIPTION
**Pull request by lukevs**
Preliminary and no support for profit yet

Spot check verified that it detects sandwiches:
https://twitter.com/bertcmiller/status/1443532075936751621/photo/1
```
INFO:mev_inspect.inspect_block:Block: 12219145 -- Found 1 sandwiches
{
    "block_number": 12219145,
    "sandwicher_address": "0x4d3e152aa1426ecb72c552bec5f2375f6327333f",
    "frontrun_transaction_hash": "0x020ec558f3da5ec292412a180c0589f3fb8748d9721d9e0dd293ef20368d9f88",
    "backrun_transaction_hash": "0xd35167f7dc9b628b2adc652692f826c767e348c8446fa2729f08e6158bfe8635",
    "sandwiched_swaps": [
        {
            "abi_name": "UniswapV2Pair",
            "transaction_hash": "0x3713bf39594648773f678cba86966e78dd010e8c1da3eb6ce73a5d8220c9df9f",
            "transaction_position": 1,
            "block_number": 12219145,
            "trace_address": [
                4
            ],
            "contract_address": "0x86fef14c27c78deaeb4349fd959caa11fc5b5d75",
            "from_address": "0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc",
            "to_address": "0x57c3606b4a9f22525d735fb49be57fbfbbc485c7",
            "token_in_address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "token_in_amount": 9259571355264578089,
            "token_out_address": "0xfca59cd816ab1ead66534d82bc21e7515ce441cf",
            "token_out_amount": 677053375760127283617,
            "protocol": null,
            "error": null
        }
    ]
}
```

https://twitter.com/willwarren/status/1365706011181481993
```
INFO:mev_inspect.inspect_block:Block: 11940345 -- Found 1 sandwiches
{
    "block_number": 11940345,
    "sandwicher_address": "0x000000000000084e91743124a982076c59f10084",
    "frontrun_transaction_hash": "0x9715c4da5ba1f4fcb8bf23d9b86c864e87a26decb129922545d75d2067990803",
    "backrun_transaction_hash": "0x19c0890b3ae612bba3dfc73d4ea00f89edcc22d748dc2510277d154366dd27e1",
    "sandwiched_swaps": [
        {
            "abi_name": "UniswapV2Pair",
            "transaction_hash": "0xe29e8a1b593d023b19d158be6cf65b691c8234440307c9fb14161ffe75f457fc",
            "transaction_position": 88,
            "block_number": 11940345,
            "trace_address": [
                3
            ],
            "contract_address": "0x2f85e11f6f12ead6af643f083a34e001030d2a6f",
            "from_address": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
            "to_address": "0x984c5d268b220784e87fbe8edbb5c6b9f7ba9fc4",
            "token_in_address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "token_in_amount": 20000000000000000000,
            "token_out_address": "0x6149c26cd2f7b5ccdb32029af817123f6e37df5b",
            "token_out_amount": 1316598702516906146186,
            "protocol": null,
            "error": null
        }
    ]
}
```

Handles multiple transactions being sandwiched at once as well
https://twitter.com/bertcmiller/status/1412566588419608577
```
{
    "block_number": 12775690,
    "sandwicher_address": "0x000000000027d2efc283613d0c3e24a8b430c4d8",
    "frontrun_transaction_hash": "0x91a3abe5f3b806426542252820ba0ab6d56c098fdef6864ecaf4d352f64217a0",
    "backrun_transaction_hash": "0xc300d1ff79d3901b58dc56489fc7d083a6c13d422bfc1425a0579379300c95a2",
    "sandwiched_swaps": [
        {
            "abi_name": "UniswapV2Pair",
            "transaction_hash": "0x9b40deca1f53593b7631ca25485d0c6faf90279b9872845acfd5c98afb185934",
            "transaction_position": 3,
            "block_number": 12775690,
            "trace_address": [
                3
            ],
            "contract_address": "0xefb47fcfcad4f96c83d4ca676842fb03ef20a477",
            "from_address": "0x03f7724180aa6b939894b5ca4314783b0b36b329",
            "to_address": "0x37e17e96736aee2ca809abd91e0f8744910ca19a",
            "token_in_address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "token_in_amount": 652974555369106606,
            "token_out_address": "0x9813037ee2218799597d83d4a5b6f3b6778218d9",
            "token_out_amount": 60000000000000000000,
            "protocol": null,
            "error": null
        },
        {
            "abi_name": "UniswapV2Pair",
            "transaction_hash": "0xf8e45a291cdab5e456375e4d7df30771670d504835c9332b32114e5bc4e315f9",
            "transaction_position": 4,
            "block_number": 12775690,
            "trace_address": [
                3
            ],
            "contract_address": "0xefb47fcfcad4f96c83d4ca676842fb03ef20a477",
            "from_address": "0x03f7724180aa6b939894b5ca4314783b0b36b329",
            "to_address": "0xd3b7ddf9eb72837f0ee3d1d30dec0e45fbdf79b1",
            "token_in_address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "token_in_amount": 300000000000000000,
            "token_out_address": "0x9813037ee2218799597d83d4a5b6f3b6778218d9",
            "token_out_amount": 27561865602394087181,
            "protocol": null,
            "error": null
        },
        {
            "abi_name": "UniswapV2Pair",
            "transaction_hash": "0xdf63b22773b66cc41e00fd42c3b3c7f42912f87476ffe6d821e3f5c00284f00b",
            "transaction_position": 5,
            "block_number": 12775690,
            "trace_address": [
                3
            ],
            "contract_address": "0xefb47fcfcad4f96c83d4ca676842fb03ef20a477",
            "from_address": "0x03f7724180aa6b939894b5ca4314783b0b36b329",
            "to_address": "0xcf99e104fdc46bea618d85ac5250067f19a56e41",
            "token_in_address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "token_in_amount": 125000000000000000,
            "token_out_address": "0x9813037ee2218799597d83d4a5b6f3b6778218d9",
            "token_out_amount": 11483313070817976324,
            "protocol": null,
            "error": null
        },
        {
            "abi_name": "UniswapV2Pair",
            "transaction_hash": "0x1fe35f66e24f12bdb54a0d35934aac809c783710d998621b70116ea9f95f4f4f",
            "transaction_position": 6,
            "block_number": 12775690,
            "trace_address": [
                3
            ],
            "contract_address": "0xefb47fcfcad4f96c83d4ca676842fb03ef20a477",
            "from_address": "0x03f7724180aa6b939894b5ca4314783b0b36b329",
            "to_address": "0xd7c9f3010efdff665ee72580ffa7b4141e56b17e",
            "token_in_address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "token_in_amount": 30000000000000000000,
            "token_out_address": "0x9813037ee2218799597d83d4a5b6f3b6778218d9",
            "token_out_amount": 2742522049933966038599,
            "protocol": null,
            "error": null
        }
    ]
}
```

### CODE REVIEW

File: mev_inspect/sandwiches.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                and other_swap.token_out_address == front_swap.token_out_address
+                and other_swap.from_address != sandwicher_address
+            ):
+                sandwiched_swaps.append(other_swap)
+            elif (
+                other_swap.token_out_address == front_swap.token_in_address
+                and other_swap.token_in_address == front_swap.token_out_address
+                and other_swap.from_address == sandwicher_address
+            ):
+                return Sandwich(
```
Comment (lukevs) [inline mev_inspect/sandwiches.py:51]: TODO 
```
if len(sandwiched_swaps) > 0
```

it's not a sandwich without fillings

**Comment by bertmiller** [comment]
Block `12753463` is an interesting test case. It's a sandwich with 4 txs in the meat that is also backrunning 2 transactions. It seems to have thrown your classifier off some, I think you classify the sandwiched transactions wrong there (one of the swaps is a backrun, one is the actual ending swap of the sandwich.

Oddly enough there are what seem to be two innocuous transactions from the same account selling stuff in this block, one of them which is being sandwiched and one that got through in the mempool later, and I think that throws your heuristics off.

All in all while this is a very interesting test case I think it's likely an edge case and not a showstopper for this PR.

**Comment by lukevs** [comment]
Broken into pieces - the main logic now lives here https://github.com/flashbots/mev-inspect-py/pull/163
